File: ein.py/src/ein/inference/lookahead.py
"""One-step contradiction lookahead — S1.5.6 T1.5.6.2.

A hypothesis-generation filter: before a candidate fact ``h`` is
emitted (and later forked + saturated by the lattice search), ask the
cheap question — *"does adding ``h`` to the already-saturated KB
produce a contradiction in a single rule firing?"*. If yes, drop
``h``; it would only fork, saturate, and die.

This is the index-based fast path the S1.5.4 "Topic B Tier B"
complaint asked for: paying for a fork + full saturation just to
learn a hypothesis dies in one rule step is wasteful.
:meth:`Lookahead.dies_immediately` runs *one* rule step instead.

This is a *hypgen-time* filter: it prunes a candidate at generation
time — before any fork or saturation — rather than discovering the
death only after forking and saturating a branch.

Mechanism
---------
The candidate ``h`` is injected into each ``:match`` premise whose
relation it unifies with; the rule's *other* premises are run
against the saturated KB through the ordinary matcher (no fork, no
mutation). For every resulting ``:assert`` fact ``f`` the engine
checks whether the would-be KB (``h`` + ``f`` added) holds a
contradiction the P1.4 ``ContradictionDetector`` would flag:

- ``f`` is ``(false)``                  — direct ⊥;
- ``f`` is ``(not h)``                  — ``h`` self-falsified;
- ``f`` is ``(not g)``, ``g`` already a REASONING-layer fact
                                        — a same-layer ``(g, ¬g)`` pair;
- ``f`` is positive and ``(not f)`` already a REASONING-layer fact
                                        — ``h`` derives a forbidden fact.

The REASONING-layer guards mirror the detector exactly: a
cross-layer ``(X, ¬X)`` (a derived ``(not g)`` against an authored
FACT-layer ``g``) is *not* a contradiction, so the lookahead must
not kill on it. The filter only ever *under*-approximates death —
a missed death just forks and dies normally; it never reports a
live hypothesis as dead.
"""
from __future__ import annotations

from ein.kb.entities import Fact, Layer
from ein.kb.store import KnowledgeBase

from . import match, primitives
from .compile import AbsentGuard, Join, JoinPlan, NafGuard, NestedPattern, Scan
from .engine import Engine
from .firing import build_fact
from .match import _bind_args
from .world import project


class Lookahead:
    """One-step contradiction simulator over a fixed rule set.

    Compile the rule plans once (``Lookahead(kb)``); reuse
    :meth:`dies_immediately` across every candidate. The compiled
    plans bake in activator-bound relation names, so a Lookahead
    built at the root KB stays valid for every fork (forks share
    the rule set).
    """

    def __init__(self, kb: KnowledgeBase) -> None:
        engine = Engine(kb)
        engine.compile_all()
        # `compile_all` walks `kb.rules` only — `(hrule …)` rules
        # live in `kb.hrules` and are never compiled here, so no
        # filtering is needed (S1.5.6b).
        self._plans: tuple[JoinPlan, ...] = tuple(engine.cache.values())

    def dies_immediately(self, kb: KnowledgeBase, h: Fact) -> bool:
        """True iff adding ``h`` to ``kb`` yields a one-step contradiction.

        ``kb`` is expected to be saturated — the engine only calls
        this on a post-saturation KB. Read-only: no fork, no
        mutation of ``kb``.
        """
        for plan in self._plans:
            # A rule may conclude SEVERAL facts (S1.8.A13) — any one could
            # contradict `h`, so probe every fact-shaped conclusion.
            fact_templates = [
                t for t in plan.assert_templates
                if isinstance(t, NestedPattern)
            ]
            if not fact_templates:
                continue
            for steps, guards in plan.disjuncts():
                if _unjudgeable(guards):
                    # S1.21.8 (D3) — a guard we cannot evaluate in the world
                    # *with* `h` must not be assumed to pass. Skipping the
                    # disjunct only loses a kill, which is the safe direction:
                    # the contract is to under-approximate death.
                    continue
                for idx, step in enumerate(steps):
                    if not isinstance(step, (Scan, Join)):
                        continue
                    if step.relation != h.relation_name:
                        continue
                    # Inject `h` into this premise — unify its args.
                    seed = _bind_args(
                        step.arg_slots, h.args, dict(plan.bindings_seed),
                    )
                    if seed is None:
                        continue
                    # Run the rule's other premises against the KB; the
                    # injected premise is supplied by `h`.
                    rest = steps[:idx] + steps[idx + 1:]
                    probe = JoinPlan(
                        rule_name=plan.rule_name,
                        activator_args=plan.activator_args,
                        bindings_seed=seed,
                        steps=rest,
                        why=plan.why,
                    )
                    for bindings, _premises in match.run(probe, kb):
                        if not _guards_pass_with(guards, bindings, kb, h):
                            continue
                        for template in fact_templates:
                            try:
                                f = build_fact(template, bindings)
                            except (KeyError, TypeError):
                                # Defensive: a malformed assert template
                                # never kills a candidate.
                                continue
                            if _is_contradiction(kb, f, h):
                                return True
        return False


def _unjudgeable(guards: tuple[NafGuard, ...]) -> bool:
    """True iff some guard's verdict in ``kb`` plus ``h`` cannot be decided cheaply.

    A guard containing a *nested* absent (what a ``forall`` desugars to) is
    non-monotone in the KB: adding ``h`` can make the inner absent fail and so
    make the outer **pass**. Deciding that needs the real world, which the
    lookahead deliberately does not build. Report unjudgeable and skip.
    """
    return any(_has_nested_absent(g.sub_steps) for g in guards)


def _has_nested_absent(steps: tuple[object, ...]) -> bool:
    return any(isinstance(st, AbsentGuard) for st in steps)


def _guards_pass_with(
    guards: tuple[NafGuard, ...],
    bindings: dict,
    kb: KnowledgeBase,
    h: Fact,
) -> bool:
    """Do ``guards`` still pass in the world ``kb`` plus ``h``?

    This is the D3 fix (P1.21 R4). The probe *hypothesises* ``h`` into one
    positive premise but the KB it runs against does not contain it, so
    evaluating a guard against ``kb`` alone answers a question about a
    different world — and a rule whose guard watches the candidate's own
    relation would kill a live hypothesis, in violation of the "never reports
    a live hypothesis as dead" contract.

    Two checks, no mutation: the guard must find no match in ``kb``, **and**
    ``h`` must not create one. `_unjudgeable` has already excluded the
    non-monotone (nested-absent) shapes, so for what remains those two
    together are exactly "no match in ``kb`` with ``h`` added".
    """
    for g in guards:
        scoped = project(bindings, g.scope)
        for _ in match.run_steps(g.sub_steps, dict(scoped), (), kb):
            return False
        for _ in match._seed_steps(g.sub_steps, scoped, h, kb):
            return False
    return True


def _is_contradiction(kb: KnowledgeBase, f: Fact, h: Fact) -> bool:
    """True iff a KB holding ``h`` + the derived ``f`` is contradictory.

    Mirrors :class:`~ein.inference.contradiction.ContradictionDetector`
    — the direct-⊥ shape and the same-layer ``(X, ¬X)`` pair. Both
    ``h`` and any rule-derived ``f`` live at the REASONING layer.
    """
    # Direct ⊥ — a `(false …)` fact.
    if f.relation_name == primitives.FALSE:
        return True
    # `f` is a negative `(not g)`.
    if f.relation_name == primitives.NOT:
        if not f.args or not isinstance(f.args[0], Fact):
            return False
        g = f.args[0]
        # `(not h)` against `h` — both REASONING, a guaranteed pair.
        if (g.relation_name, g.args) == (h.relation_name, h.args):
            return True
        # `(not g)` against an existing positive `g` — a pair only
        # when `g` is itself REASONING-layer (cross-layer is not a
        # contradiction; see the detector's module docstring).
        existing = kb._fact_by_id(g.relation_name, g.args)
        return existing is not None and existing.layer is Layer.REASONING
    # `f` is positive — `h` would derive it. A contradiction iff
    # `(not f)` already exists at the REASONING layer.
    if (f.relation_name, f.args) not in kb._negated_facts:
        return False
    neg = kb._fact_by_id(primitives.NOT, (f,))
    return neg is not None and neg.layer is Layer.REASONING


__all__ = ["Lookahead"]

File: ein.rs/crates/ein-infer/src/lookahead.rs
//! One-step contradiction lookahead — the costliest hypgen filter.
//!
//! Before a candidate `h` is emitted (and later forked and saturated), ask the
//! cheap question: *does adding `h` to the already-saturated KB produce a
//! contradiction in a single rule firing?* If yes, drop it — it would only
//! fork, saturate and die.
//!
//! The mechanism is the matcher's own semi-naive seed. `h` is injected into
//! each `:match` premise whose relation it unifies with, the rule's *other*
//! premises run against the saturated KB, and every resulting `:assert` fact
//! is checked against the detector's shapes. No fork, no saturation, and the
//! only write is the kill cache one frame out ([`crate::hypgen`]).
//!
//! ### The contract, and why it is one-sided
//!
//! The filter may only ever **under**-approximate death: a missed kill just
//! forks and dies normally, but a hypothesis wrongly reported dead is silently
//! lost. Two guards keep it on the safe side and both are semantics, not
//! optimisations:
//!
//! - a disjunct whose guards cannot be judged in the world *with* `h` is
//!   skipped rather than assumed to pass ([`unjudgeable`]);
//! - a guard that can be judged is judged against `kb` **and** against the
//!   one fact `h` adds ([`Lookahead::guards_pass_with`]) — the probe
//!   hypothesises `h` into a positive premise but the KB it runs against does
//!   not contain it, so asking `kb` alone would answer about a different
//!   world, and a rule whose guard watches the candidate's own relation would
//!   kill a live hypothesis.
//!
//! ### Its sign
//!
//! `enable_pre_branch_lookahead` measures *slightly negative* today (0.9× on
//! exhaustive `zebra2`): it pays a simulation to avoid forks that S1.9.E23's
//! fail-fast already made cheap. Because it runs the matcher it inherits
//! [design/05](../../../../plans/m1a_rust/design/05_matcher.md)'s speedup
//! wholesale, which may flip that sign — a
//! [P1a.6](../../../../plans/m1a_rust/p1a.6_performance/README.md)
//! measurement, not a decision to take here. The default does not move
//! without it.

use std::ops::ControlFlow;
use std::sync::Arc;

use ein_core::{FactId, Kb, Terms, Value};

use crate::compile::CompileError;
use crate::firing::{Env, build_fact};
use crate::match_::Matcher;
use crate::plan::{NafGuard, Plan, Span, Step};
use crate::saturator::Session;

/// A one-step contradiction simulator over a fixed rule set.
///
/// Compile the plans once, reuse [`Lookahead::dies_immediately`] across every
/// candidate. The plans bake in activator-bound relation names, so one built
/// at the root stays valid for every fork — forks share the rule set.
pub struct Lookahead {
    /// `compile_all` walks `kb.rules` only, so an `(hrule …)` is never in
    /// here and no filtering is needed: the lookahead cannot simulate a
    /// speculation.
    plans: Vec<Arc<Plan>>,
}

impl Lookahead {
    pub fn new(s: &mut Session<'_>) -> Result<Lookahead, CompileError> {
        let mut engine = crate::engine::Engine::with_memo(s.memo.clone());
        engine.compile_all(s.ast, s.terms, s.kb, s.events)?;
        let plans = (0..engine.len()).map(|i| engine.plan_arc(i)).collect();
        Ok(Lookahead { plans })
    }

    /// True iff adding `h` to `kb` yields a one-step contradiction.
    ///
    /// `kb` is expected to be saturated — the engine only calls this on a
    /// post-saturation KB. Read-only: no fork, and no mutation of the KB.
    pub fn dies_immediately(&self, s: &mut Session<'_>, m: &mut Matcher, h: FactId) -> bool {
        ein_core::counters::bump(|c| c.lookahead_probe += 1);
        let rel = s.terms.facts.rel(h);
        let mut envs: Vec<(Vec<Value>, Vec<crate::plan::Reg>)> = Vec::new();
        for plan in &self.plans {
            // A rule may conclude several facts (S1.8.A13) and any one could
            // contradict `h`, so every fact-shaped conclusion is probed.
            let templates: Vec<crate::plan::Slot> = plan
                .asserts
                .iter()
                .copied()
                .filter(|t| matches!(t, crate::plan::Slot::Nested { .. }))
                .collect();
            if templates.is_empty() {
                continue;
            }
            for (i, d) in plan.disjuncts.iter().enumerate() {
                if unjudgeable(plan, *d) {
                    // S1.21.8 (D3) — a guard we cannot evaluate in the world
                    // *with* `h` must not be assumed to pass. Skipping the
                    // disjunct only loses a kill, which is the safe direction.
                    continue;
                }
                let guards = plan.guards(d.guards);
                for (at, step) in plan.steps(d.steps).iter().enumerate() {
                    let Step::Rel(r) = step else { continue };
                    if r.rel != rel {
                        continue;
                    }
                    // The matcher borrows `terms` immutably and `build_fact`
                    // interns, so one seeding's environments are collected and
                    // its conclusions built after the walk. Sound here for a
                    // stronger reason than in `Hrules::candidates`: this walk
                    // writes nothing at all, to the KB or anywhere else.
                    envs.clear();
                    m.run_seeded_at(s.kb, s.terms, s.ast, plan, i, at, h, &mut |mt| {
                        envs.push((mt.regs().to_vec(), mt.trail().to_vec()));
                        ControlFlow::Continue(())
                    });
                    for (regs, trail) in &envs {
                        if !self.guards_pass_with(s, m, plan, guards, regs, h) {
                            continue;
                        }
                        for &template in &templates {
                            let env = Env {
                                regs,
                                trail,
                                premises: &[],
                            };
                            // Defensive, as in ein.py: a malformed assert
                            // template never kills a candidate.
                            let Ok(f) = build_fact(s.terms, plan, env, template) else {
                                continue;
                            };
                            if is_contradiction(s.kb, s.terms, f, h) {
                                return true;
                            }
                        }
                    }
                }
            }
        }
        false
    }

    /// Do `guards` still pass in the world `kb` **plus** `h`?
    ///
    /// Two checks, no mutation: the guard must find no match in `kb`, and `h`
    /// must not create one. [`unjudgeable`] has already excluded the
    /// non-monotone shapes, so for what remains those two together are exactly
    /// "no match in `kb` with `h` added".
    fn guards_pass_with(
        &self,
        s: &Session<'_>,
        m: &mut Matcher,
        plan: &Plan,
        guards: &[NafGuard],
        regs: &[Value],
        h: FactId,
    ) -> bool {
        for g in guards {
            if m.holds(s.kb, s.terms, s.ast, plan, g, regs) {
                return false;
            }
            if m.holds_seeded(s.kb, s.terms, s.ast, plan, g, regs, h) {
                return false;
            }
        }
        true
    }
}

/// True iff some guard's verdict in `kb` plus `h` cannot be decided cheaply.
///
/// A guard containing a *nested* absent — what a `forall` desugars to — is
/// non-monotone in the KB: adding `h` can make the inner absent fail and so
/// make the outer **pass**. Deciding that needs the real world, which the
/// lookahead deliberately does not build.
fn unjudgeable(plan: &Plan, d: crate::plan::Disjunct) -> bool {
    plan.guards(d.guards)
        .iter()
        .any(|g| has_nested_absent(plan, g.sub))
}

fn has_nested_absent(plan: &Plan, sub: Span) -> bool {
    plan.steps(sub)
        .iter()
        .any(|st| matches!(st, Step::Absent { .. }))
}

/// True iff a KB holding `h` **and** the derived `f` is contradictory.
///
/// Mirrors [`crate::contradiction`]'s two shapes — direct ⊥ and the `(X, ¬X)`
/// pair, which since S1.22.1b is a contradiction regardless of how either side
/// entered the KB. Its semantics are the *hypothetical* KB's, which is why it
/// is not `contradicts`: `h` is not in `kb`.
fn is_contradiction(kb: &Kb, terms: &Terms, f: FactId, h: FactId) -> bool {
    let (rel, args) = terms.facts.get(f);
    // Direct ⊥ — a `(false …)` fact.
    if rel == terms.kernel.r#false {
        return true;
    }
    if rel == terms.kernel.not {
        // `f` is a negative `(not g)`.
        let Some(g) = args.first().and_then(|a| a.as_fact()) else {
            return false;
        };
        // `(not h)` against `h` — both hypothetical, a guaranteed pair.
        if g == h {
            return true;
        }
        // `(not g)` against an existing positive `g` — a pair, whatever `g`'s
        // origin.
        return kb.contains(g);
    }
    // `f` is positive — `h` would derive it. A contradiction iff `(not f)` is
    // already in the KB.
    kb.is_negated(f)
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::compile::SharedMemo;
    use ein_core::Kb;
    use ein_ir::{Ast, from_ir::load, parse};

    /// Saturate `src`, then ask whether `(rel arg)` dies immediately.
    fn dies(src: &str, rel: &str, arg: &str) -> bool {
        let mut ast = Ast::new();
        let mut terms = Terms::new();
        let forms = parse(&mut ast, src, Some("<test>")).expect("parses");
        let mut kb: Kb = load(&mut ast, &mut terms, &forms, None).expect("loads");
        let mut ev = crate::events::Events::off();
        let mut s = Session {
            kb: &mut kb,
            terms: &mut terms,
            ast: &ast,
            events: &mut ev,
            memo: SharedMemo::default(),
        };
        let mut sat = crate::saturator::Saturator::new(&mut s).expect("compiles");
        sat.saturate(&mut s, None, &mut |_| {}).expect("saturates");
        let r = s.terms.intern_text(rel).expect("room");
        let a = s.terms.intern_text(arg).expect("room");
        let h = s.terms.intern_fact(r, &[Value::sym(a)]).expect("room");
        let l = Lookahead::new(&mut s).expect("compiles");
        l.dies_immediately(&mut s, &mut Matcher::new(), h)
    }

    /// P1.21 R4 — the guard is judged in the world `kb` **plus** `h`.
    ///
    /// `false ← (cand ?x) ∧ (absent (cand ?x))` can never fire in any real
    /// match: its premises are jointly unsatisfiable in one world. The probe
    /// nonetheless posits `h` into the positive premise, so evaluating the
    /// guard against a KB *without* `h` would answer about a different world
    /// and report a live hypothesis dead — the one thing the filter may never
    /// do. `examples/branching/13_lookahead_naf_world.ein` is the same shape
    /// as a whole-solve fixture.
    #[test]
    fn the_naf_world_includes_the_candidate() {
        let src = "(rule self-block ()\n  :match (and (cand ?x) (absent (cand ?x)))\n\
                   \x20 :assert (false)\n  :why \"unsatisfiable in any one world\" :priority 100)\n\
                   (relation cand T)";
        assert!(!dies(src, "cand", "A"));
    }

    /// A **nested** absent is the shape the filter must refuse to judge.
    ///
    /// The guard reads "no `p` that is not blocked by an unmet `q`". Adding
    /// `(s B)` makes the *inner* absent fail, which makes the middle one
    /// pass, which gives the outer query a match — so the guard fails in
    /// `kb + h` and the rule does not fire. Neither cheap check sees that:
    /// `kb` alone has no match, and `h` is not a positive premise of the
    /// query, so seeding it finds nothing either. Judged by those two the
    /// guard would "pass", `(false)` would be derived, and a live hypothesis
    /// would be lost — which is why the whole disjunct is skipped instead.
    /// `examples/branching/14_lookahead_unjudgeable.ein` is the same shape as
    /// a whole-solve fixture.
    #[test]
    fn a_nested_absent_is_not_judged_at_all() {
        let src = "(rule r ()\n  :match (and (s ?x) (absent (and (p ?y) \
                   (absent (and (q ?y) (absent (s ?y)))))))\n  :assert (false) \
                   :priority 100)\n\
                   (relation s T) (relation p T) (relation q T)\n\
                   (p B :source \"(1)\") (q B :source \"(2)\")";
        assert!(!dies(src, "s", "B"));
    }

    /// …and the fix must not disarm the filter: a purely positive rule that
    /// derives `(false)` from the candidate still kills it, and only for the
    /// argument that has the other premise.
    #[test]
    fn a_positive_rule_still_kills() {
        let src = "(rule blow-up ()\n  :match (and (cand ?x) (bad ?x))\n  :assert (false)\n\
                   \x20 :why \"cand + bad is absurd\" :priority 100)\n\
                   (relation cand T)\n(relation bad T)\n(bad A :source \"(1)\")";
        assert!(dies(src, "cand", "A"));
        assert!(!dies(src, "cand", "B"));
    }
}
